File: programming/basics/crackingthecodinginterview/linked_lists/3_main.py
class Node:
    def __init__(self, val=None, next=None):
        self.val = val
        self.next = next

def print_ll(head: Node):
    current = head
    while current:
        print(current.val)
        current = current.next

def remove_middle(n: Node):
    n.val = n.next.val
    n.next = n.next.next
    n.next = None

head = Node(9)
a = Node(9)
b = Node(7)
c = Node(9)

head.next = a
a.next = b
b.next = c

print_ll(head)
print("---")
remove_middle(b)
print_ll(head)

File: programming/basics/crackingthecodinginterview/linked_lists/3_main.c
// delete middle node
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node *next;
};

typedef struct Node Node;

void remove_middle(Node *n)
{
    n->data = n->next->data;
    n->next = n->next->next;
    free(n->next);
}

void print_ll(Node *head) {
    Node *current = head;
    while (current != NULL) {
        printf("%d\n", current->data);
        current = current->next;
    }
}

int main()
{
    Node *head = (Node *)malloc(sizeof(Node));
    head->data = 1;
    head->next = (Node *)malloc(sizeof(Node));

    head->next->data = 2;
    head->next->next = (Node *)malloc(sizeof(Node));

    head->next->next->data = 3;
    head->next->next->next = NULL;

    print_ll(head);
    printf("---\n");
    remove_middle(head->next);
    print_ll(head);

    return 0;
}
